### DISCUSSION TITLE
Quarkus Testing - JVM/Native mode

### DISCUSSION THREAD

**Discussion by gdcmarinho** [quarkusio/quarkus#24954] [Q&A]
I've been studying about Quarkus for some  weeks and I'm currently studying about testing in Quarkus. The docs reference to this topic  that I'm using is [Testing your application](https://quarkus.io/guides/getting-started-testing) and in the index of the documentation there is a reference about how to test using this different types of modes (JVM/Native), but I didn't get what is the real difference between these modes. So I'd like to ask for someone with more experience: Testing in JVM and Native mode are complementary ways to test applications or they're two different paths to test the same kind of tests?

**Comment by gdcmarinho** [comment]
@Sanne can you help me with this information?

**Reply by Sanne** [reply]
You should always start to test your application in JVM mode, as:

1. it's easier & faster to compile an application
2. there are better debugging tools
3. If it doesn't work correctly in JVM mode it's not going to work in native either, so consider this a pre-requisite.

But when everything works as expected in JVM, native mode might have additional requirements; best to refer to https://www.graalvm.org/22.0/reference-manual/native-image/ to have an idea of these requirements - but don't get intimidated as Quarkus takes care of most requirements, especially those for libraries and other code you don't control yourself.

So to ensure the application also works in native, the produced binary should also be tested. Testing methodology might be slightly different, as we prefer to test it as a black box rather than running unit tests; so the testing helpers provided by Quarkus (and mentioned in the documentation you linked) will start the native binary for you and then typically you'd invoke its rest APIs (when building a REST server) to verify it's responding as expected, or connect to it in any other way if building a different type of application.
